src/js/stores/PluginStore.js

```diff
@@ -1,78 +1,121 @@
 import {EventEmitter} from "events";
 
 import AppDispatcher from "../AppDispatcher";
-import PluginDispatcher from "../plugin/external/PluginDispatcher";
+import Util from "../helpers/Util";
 import DialogActions from "../actions/DialogActions";
 import DialogSeverity from "../constants/DialogSeverity";
 import PluginActions from "../actions/PluginActions";
 import PluginEvents from "../events/PluginEvents";
+import PluginModules from "../constants/PluginModules";
+import States from "../constants/States";
+import pluginScheme from "../stores/schemes/pluginScheme";
 
-import Util from "../helpers/Util";
+var plugins = [];
+
+function addPlugin(data) {
+  if (data.id == null && !Util.isObject(data.info)
+    && getPluginById(data.id) != null) {
+    return;
+  }
 
-var pluginsMetaData = [];
-var pluginsToLoad = [];
-var pluginsStarted = [];
-var pluginsErrored = [];
+  plugins.push(Util.extendObject(pluginScheme, data.info, {id: data.id}));
+}
 
-var bootstrapComplete = false;
+function getPluginById(id) {
+  return plugins.find((plugin) => plugin.id === id);
+}
 
-function loadNextPlugin() {
-  if (pluginsToLoad.length === 0) {
+function updatePluginState(id, state) {
+  var plugin = getPluginById(id);
+  if (plugin == null) {
     return;
   }
-  PluginActions.requestPlugin(pluginsToLoad.shift());
+
+  plugin.state = state;
 }
 
-function checkForStartupCompleteAndEmit() {
-  let allStarted = pluginsMetaData.every(plugin => {
-    return pluginsStarted.includes(plugin.id) ||
-      pluginsErrored.some(errored => errored.id === plugin.id);
+function loadPlugins() {
+  plugins.filter((plugin) => {
+    return plugin.modules.indexOf(PluginModules.UI) >= 0
```

Comment:
Brandnew: You can use `.includes(value)` in favor of `.indexOf()`. It returns a boolean.

Correct function without unnecessary parentheses, corrected indentation and .includes().
```JavaScript
function loadPlugins() {
  plugins
    .filter(plugin => {
      return plugin.modules.includes(PluginModules.UI)
        && plugin.state === States.STATE_INITIAL;
    })
    .forEach(plugin => {
      plugin.state = States.STATE_LOADING;
      PluginActions.loadPlugin(plugin.id);
    });
}
```